plugins/outputs/kafka/kafka.go

```diff
@@ -235,30 +256,69 @@ func (k *Kafka) Description() string {
 }
 
 func (k *Kafka) Write(metrics []telegraf.Metric) error {
+	var (
+		metricsmap  = map[string]map[string][]telegraf.Metric{}
+		routingTags = map[string]bool{}
+		routingTag  string
+	)
+
 	if len(metrics) == 0 {
 		return nil
 	}
 
 	for _, metric := range metrics {
-		buf, err := k.serializer.Serialize(metric)
-		if err != nil {
-			return err
+		routingTag = metric.Tags()[k.RoutingTag]
+		if _, found := routingTags[routingTag]; !found {
+			metricsmap[routingTag] = make(map[string][]telegraf.Metric)
 		}
+		routingTags[routingTag] = true
```

Comment:
I think you can remove the `routingTags` map, and just base the creation of the submap on if `_, found := routingTags[routingTag]; !found` is true.